# Construir sua primeira rede neural
A primeira etapa é definir o vetor de entrada

In [39]:
input_vector = [1.72, 1.23]

A segunda consiste em escolher os pesos

In [40]:
weights_1 = [1.26, 0]
weights_2 = [2.17, 0.32]

Calcular o produto interno do vetor de entrada pelo pesos 1

In [41]:
first_indexes_mult = input_vector[0] * weights_1[0]
second_indexes_mult = input_vector[1] * weights_1[1]
dot_product_1 = first_indexes_mult + second_indexes_mult

In [63]:
print(f"O produto da primeira entrada pelo peseo é: {first_indexes_mult}")

O produto da primeira entrada pelo peseo é: 2.1672


In [65]:
print(f"O produto da primeira entrada pelo peseo é: {second_indexes_mult}")

O produto da primeira entrada pelo peseo é: 0.0


In [66]:
print(f"O produto interno é: {dot_product_1}")

O produto interno é: 2.1672


### Repetir as operações usando numpy
Importar a biblioteca numpy

In [67]:
import numpy as np

Calcular o produto interno do vetor de entrada pelo pesos 1 usando numpy

In [47]:
dot_product_1 = np.dot(input_vector, weights_1)

In [48]:
print(f"O produto interno é: {dot_product_1}")

O produto interno é: 2.1672


Calcular o produto interno do vetor de entrada pelo pesos 2 usando numpy


In [49]:
dot_product_2 = np.dot(input_vector, weights_2)

In [50]:
print(f"O produto interno é: {dot_product_2}")

O produto interno é: 4.1259999999999994


# Um problema de classificação

| Vetor de Entrada | Alvo |
| :- | -: |
| [1.66, 1.56] | 1 |
|   [2, 1.5]   | 0 |

#### Envolvendo os vetores em matrizes NumPy

In [51]:
input_vector = np.array([1.66, 1.56])
weights_1 = np.array([1.45, -0.66])
bias = np.array([0.0])

#### Definição da função sigmoide

In [52]:
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

#### Definição da função de predição

In [53]:
def make_prediction(input_vector, weights, bias):
     layer_1 = np.dot(input_vector, weights) + bias
     layer_2 = sigmoid(layer_1)
     return layer_2

In [54]:
prediction = make_prediction(input_vector, weights_1, bias)
print(f"O resultado da predição é: {prediction}")

O resultado da predição é: [0.7985731]


Modificando o vetor de entrada

In [55]:
input_vector = np.array([2, 1.5])

In [56]:
prediction = make_prediction(input_vector, weights_1, bias)
print(f"O resultado da predição é: {prediction}")

O resultado da predição é: [0.87101915]


| Vetor de Entrada | Alvo | Predito |
| :- | -: | -: |
| [1.66, 1.56] | 1 | 0.798 |
|   [2, 1.5]   | 0 | 0.871 |

Treinar sua primeira rede neural:

In [57]:
target = 0
mse = np.square(prediction - target)
print(f"Predição: {prediction}; Erro: {mse}")

Predição: [0.87101915]; Erro: [0.75867436]


Calcular da derivada:

In [58]:
derivative = 2 * (prediction - target)
print(f"A derivada é {derivative}")

A derivada é [1.7420383]


Atualizar pesos:

In [68]:
print(f"O peso anterior é {weights_1}")
weights_1 = weights_1 - derivative
print(f"O peso atualizado é {weights_1}")

O peso anterior é [-2.03407659 -4.14407659]
O peso atualizado é [-3.77611489 -5.88611489]


Fazendo uma nova predição:

In [69]:
prediction = make_prediction(input_vector, weights_1, bias)
error = (prediction - target) ** 2
print(f"Prediction: {prediction}; Error: {error}")

Prediction: [7.6850629e-08]; Error: [5.90601918e-15]
